docs/tls-authentication.md

```diff
@@ -0,0 +1,248 @@
+
+  * [Generating Certificates](#generating-certificates)
+    + [Default (localhost)](#quick-and-easy)
+    + [Personalized (custom domain)](#step-by-step)
+
+  * [Using Certificates](#using-certificates)
+    + [Setting Up Cardano-SL Backend](#setting-up-cardano-sl-backend)
+    + [Contacting Cardano-SL Backend](#contacting-cardano-sl-backend)
+
+
+## Motivation and Design
+
+Communication between _daedalus_ and _cardano-sl_ is done via http server. There are various
+attacks possible to this kind of environment. In order to secure the connection between these
+two services, we rely on TLS and X.509 public key certificates signed by a known authority. 
+
+Hence, the backend server expects clients to provide valid X.509 v3 certificates signed by an
+approved authority (identified by another certificate, provided on initialization) to verify
+their identity. On the other hand, clients must check the certificate provided by the server to
+make sure they're dealing with the expected party and not a malicious service trying to hijack
+the backend's identity. 
+
+Certificates (and corresponding private keys) are generated off-band, and signed by a same CA's
+(e.g. _Certificate Authority_) private key. The steps below describe how to securely generate
+them. Securing the access to the provided private keys is of vitale importance in order to keep
+the system safe. **Any process or machine with access to the private keys may be able to
+usurp daedalus or cardano-sl's identities**.
+
+
+## Prerequisites
+
+- Install and configure [OpenSSL](https://www.openssl.org/)
+
+
+## Generating Certificates
+
+> **NOTE**: If you're setting up backend nodes and a frontend wallet on a same machine, 
+> you can ignore the _personalized_ installation and fully rely on the _default_ case.
+
+The following steps are described as if executed from a _Unix_ system (Linux or Mac OS) however,
+scripts used are also available on _Windows_ and steps could be easily transposed to this
+system.
+
+### Default (localhost, 3 years)
+
+#### Downloading Script
+
+ _[daedalus](https://github.com/input-output-hk/daedalus)_ provides a script
+ `installers/build-certificates-unix.sh` (or a windows equivalent
+ `installers/build-certificates-win64.bat`) to generate default certificates for both
+ _cardano-sl_ and _daedalus_, as well as a CA public certificate used to sign the formers. 
+
+Make sure you have an Internet access and download the installer script and the necessary
+configuration files as such:
+
+```
+$ wget -q https://raw.githubusercontent.com/input-output-hk/daedalus/master/installers/build-certificates-unix.sh
+$ wget -q https://raw.githubusercontent.com/input-output-hk/daedalus/master/installers/{ca,server,client}.conf
+```
+
+#### Running Script
+
+Run the script to generate the certificates as follows:
+
+```
+$ bash ./build-certificates-unix.sh 
+```
+
+Once done (`Oll Korrect`), a folder `tls` should have been created in your current directory
```

Comment:
One thing to note. I would assume we do, but do we handle intermediate certificates in a server.crt or client.crt file correctly? I could definitely see an organization using their existing CA (maybe in Active Directory) signing an intermediate signing authority cert/key specific for exchange API, and then issuing server/client certificates from that intermediate CA. That's what I would do if I ran an exchange at least.